Search improvements

rtibbles opened the pull request:
### Summary
* Make search a separate list endpoint.
* Return additional metadata with search results to allow filtering by channel_ids, and kinds
* Return total available search results
* Generalize the unused pagination logic in api-resource.js to allow this arbitrary metadata storage.
* Add 'view more' to search results to allow loading of additional search results, 30 at a time.
* Add 'update search' button to trigger new search based on change search string or altered filters
* Add disabled state for simple link KButton

### Reviewer guidance
Can you still search?
Can you find things you want?
Can you filter and still find things?

![image](https://user-images.githubusercontent.com/1680573/44546210-1b846200-a6cc-11e8-813f-e41732dc8f5c.png)

![image](https://user-images.githubusercontent.com/1680573/44546229-250dca00-a6cc-11e8-9b09-51b4b93d40ac.png)

![image](https://user-images.githubusercontent.com/1680573/44546246-3525a980-a6cc-11e8-8d4d-f21436026596.png)


----

### Contributor Checklist

- [x] Contributor has fully tested the PR manually
- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [x] If there are any front-end changes, before/after screenshots are included
- [x] If this is an important user-facing change, PR or related issue has a 'changelog' label

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

rtibbles committed:
Fix required GET param decorator handling.

rtibbles committed:
Do a debug log on any errors reported for devs.
Coerce Error objects properly to show them properly in the modal.

rtibbles committed:
Change content search to its own endpoint.
Add additional metadata to search for better filtering.

rtibbles committed:
Add disabled state styling for simple link.

rtibbles committed:
Add an update search button, conditionally enabled.

rtibbles commented:
@carinediaz would appreciate input on the new strings introduced here.

rtibbles committed:
Fix edge case and test behaviour.

carinediaz commented:
@rtibbles `view more` and `update search` make sense to me. I think they're both terms most people should understand and translate.

ralphiee22 commented on kolibri/core/content/api.py:
```diff
@@ -553,6 +497,100 @@ def resume(self, request, **kwargs):
+            if len(results) >= MAX_RESULTS:
+                break
+
+        total_results = len(set(queryset.filter(union(all_queries)).values_list('content_id', flat=True)))
```
does `queryset.filter(union(all_queries)).values_list('content_id', flat=True).distinct().count()` not give the correct results?

rtibbles replied:
Oh, good catch. I am always wary of using `distinct` in SQLite, but it's fine when we've already selected the `content_id`.

rtibbles replied:
Can do similar distinct calls for the two below too.

rtibbles committed:
Use more efficient queries.

ralphiee22 reviewed (approved):
👍

rtibbles committed:
Add minimal guarantees of behaviour to new search endpoint properties.

rtibbles commented:
Note: holding off on merge until frontend review has also happened.

christianmemije commented on kolibri/core/assets/src/views/buttons-and-links/buttons.scss:
```diff
@@ -152,4 +152,8 @@ a.button {
   &:hover:focus {
     outline: $core-outline;
   }
+  &:disabled {
```
Let's add this disabled state to button examples in the style guide.

rtibbles replied:
![image](https://user-images.githubusercontent.com/1680573/44688917-9c17cb00-aa0a-11e8-929c-796017f8b97e.png)
?

christianmemije commented on kolibri/plugins/learn/assets/src/views/SearchBox.vue:
```diff
@@ -106,9 +227,18 @@
       },
       search() {
         if (this.searchQuery !== '') {
```
Let's add .trim to the searchQuery v-model.

christianmemije commented on kolibri/plugins/learn/assets/src/views/SearchBox.vue:
```diff
@@ -194,4 +323,25 @@
+  }
+
+  .filter:nth-of-type(1) {
+    margin-right: 32px;
```
Let's reduce this to 16px. Let's also add bottom margins to the filters to reduce crowding. Maybe 16px? 
![localhost_8000_learn_](https://user-images.githubusercontent.com/7193975/44686903-8b645680-aa04-11e8-87eb-5e480a174144.png)

christianmemije replied:
Also, keep the vertical alignment of the icons to baseline?

rtibbles replied:
![image](https://user-images.githubusercontent.com/1680573/44689856-c74fe980-aa0d-11e8-9316-ea18dd7554ad.png)

christianmemije commented on kolibri/plugins/learn/assets/src/views/SearchPage.vue:
```diff
@@ -4,25 +4,36 @@
-      <h1 class="search-results">{{ $tr('showingResultsFor', { searchTerm }) }}</h1>
+      <h1 class="search-results">{{ $tr('showingResultsFor', {
+        searchTerm,
+          results: contents.length,
```
Weird indent?

rtibbles replied:
I think this got auto-linted this way, I did not have the additional indent when I manually edited.

rtibbles replied:
Confirmed, the linter says no to non-weird indent.

christianmemije commented on kolibri/plugins/learn/assets/src/views/SearchPage.vue:
```diff
@@ -43,9 +55,10 @@
       searchPageHeader: 'Search',
       noSearch: 'Search by typing in the box above',
-      showingResultsFor: "Results for '{searchTerm}'",
+      showingResultsFor: "Results for '{searchTerm}' ({results} of {totalResults})",
```
We should not show the `({results} of {totalResults})` part of this string when there are no results. Also would this part of the string be better off being on it's own below the heading? Also, maybe `1-30 of 70` instead of `30 of 70`?

rtibbles replied:
We don't show this string when there are no results, there is a separate `noResults` string that gets displayed instead, so think this should be fine as is.

christianmemije replied:
The heading is displayed as long as there is a search query.

rtibbles replied:
Oh, sorry, misread the template logic. We should probably make this a v-else block.

rtibbles replied:
Have updated.

rtibbles replied:
Have updated.

christianmemije reviewed (changes requested):
Definitely improved search. :1st_place_medal:  Left some minor comments.

rtibbles committed:
Update style guide for simple text button disabled state and search styling.

rtibbles commented:
Updated.

rtibbles committed:
Only show search results header if there are search results.

christianmemije commented:
Spoke to @khangmach and we decided on a few changes. 
* Instead of having an update search button, let's auto trigger a new search upon the selection of a new filter option.
* Let's change the view more button to a raised secondary to make it more prominent. 
* Let's move the num of total results to below the heading `N results`

rtibbles committed:
Frontend search tweaks.

rtibbles committed:
Set to default pointer on mouseover of disabled k-select component.

rtibbles committed:
Reset filters if a new search term is searched for.
Disable filter selection when nothing to filter by.

rtibbles committed:
Prevent repeated kinds and channel_ids in search results.

rtibbles commented:
Updated search behaviour:
![search_filter](https://user-images.githubusercontent.com/1680573/44755774-68a57100-aadc-11e8-9845-8e361c2b7556.gif)

Updated text:
![image](https://user-images.githubusercontent.com/1680573/44755796-7eb33180-aadc-11e8-9a8a-401115d41de0.png)

Updated view more:
![image](https://user-images.githubusercontent.com/1680573/44755816-9094d480-aadc-11e8-9098-8115de07a975.png)

Updated empty state:
![image](https://user-images.githubusercontent.com/1680573/44755825-a0141d80-aadc-11e8-8b82-28be3f52b528.png)
(note that this also includes an update to k-select that sets the pointer to default when disabled)

christianmemije commented on kolibri/plugins/learn/assets/src/views/SearchPage.vue:
```diff
@@ -59,7 +57,7 @@
       searchPageHeader: 'Search',
       noSearch: 'Search by typing in the box above',
-      showingResultsFor: "Results for '{searchTerm}' ({results} of {totalResults})",
+      showingResultsFor: "{totalResults} results for '{searchTerm}'",
```
Need to use the plural syntax.

rtibbles replied:
So near, yet so fars.

rtibbles replied:
Updated.

rtibbles committed:
Update string to icu plural syntax.

rtibbles committed:
Fix merge induced repeated code.
Prevent navigation induced watch triggering.

rtibbles committed:
Don't use v-model on kselect in searchbox.

christianmemije commented on kolibri/core/assets/src/views/KSelect.vue:
```diff
@@ -133,6 +133,7 @@
     methods: {
       handleChange(newSelection) {
         this.selection = newSelection;
+        this.$emit('change', newSelection);
```
Let's document this event for the style guide.

christianmemije reviewed (approved):
Left one comment. Tested and working nicely.

rtibbles committed:
Remove erroneous double change emit.
